Refactor of LDAP backend to be more efficient

john-westcott-iv opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Need engineering and QE testing badly

<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
- Bug, Docs Fix or other nominal change

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```

```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

john-westcott-iv committed:
Refactor of LDAP backend to be more efficent

kdelee commented on awx/sso/backends.py:
```diff
@@ -423,3 +457,28 @@ def on_populate_user(sender, **kwargs):
+    # Get all of the orgs listed in LDAP from the database
+    roles = ['admin_role', 'auditor_role', 'member_role']
+    prefetch_roles = ["{}__members".format(role_name) for role_name in roles]
+    orgs_tied_to_ldap = Organization.objects.filter(name__in=desired_org_states.keys()).prefetch_related(*prefecth_roles)
```
this saves us a bunch of queries from when we do   `getattr(org, role).members`

kdelee commented on awx/sso/backends.py:
```diff
@@ -423,3 +457,28 @@ def on_populate_user(sender, **kwargs):
+                getattr(org, role).members.remove(user)
+
+    # Get all of the teams listed in LDAP from the database
+    teams_tied_to_ldap = Team.objects.filter(name__in=desired_team_states.keys()).prefetch_related('member_role__members')
```
same thing here, doing the prefetch saves us a bunch of queries on ` team.member_role.members`

kdelee commented on awx/sso/backends.py:
```diff
@@ -383,31 +378,70 @@ def on_populate_user(sender, **kwargs):
+    # Get all of the orgs in the DB by name and create any new org defined in LDAP
+
+    # IS THIS ACTUALLY GETTING JUST NAMES OR DOES IT GET ALL AND THEN GET NAMES FORM THEM ALL?
+    existing_org_names = list(Organization.objects.all().values_list('name', flat=True))
```
might be able to do `Organization.objects.all().only('name')` though tbh I'm not 100% on difference

john-westcott-iv replied:
I checked and this is doing a query specifically for 'name' field:
```
tools_postgres_1    | 2022-09-24 13:41:19.087 UTC [111] LOG:  statement: SELECT "main_organization"."name" FROM "main_organization" ORDER BY "main_organization"."name" ASC
```

AlanCoding replied:
The only difference is that `.only()` still builds the python model objects with incomplete fields. If you're only going to use the name strings, then I think values_list is preferable.

john-westcott-iv committed:
Fixing bug, updating comments and adding debugging logging

AlanCoding commented on awx/sso/backends.py:
```diff
@@ -423,3 +460,34 @@ def on_populate_user(sender, **kwargs):
+    # Get all of the orgs listed in LDAP from the database
+    roles = ['admin_role', 'auditor_role', 'member_role']
+    prefetch_roles = ["{}__members".format(role_name) for role_name in roles]
+    orgs_tied_to_ldap = Organization.objects.filter(name__in=desired_org_states.keys()).prefetch_related(*prefetch_roles)
```
I realize that this is may be un-fixable in the abstract due to the business requirements, we should keep in mind that this is a heavy `prefetch_related` that could, at some point, introduce problems due to the quantity of data that it returns. The number of items it returns scales with the total number of admin/auditor/members in all the filtered organizations. But... these organizations are defined by the LDAP mapping.

You're still fetching more than what you technically need, as using a custom `Prefetch` object could avoid returning memberships for users other than the current user. That would be better handled as an annotate, since the expectations of prefetch get difficult to follow.

AlanCoding commented on awx/sso/backends.py:
```diff
@@ -383,31 +378,73 @@ def on_populate_user(sender, **kwargs):
 
-    # Update team membership based on group memberships.
-    team_map = getattr(backend.settings, 'TEAM_MAP', {})
+        # If everything returned None (because there was no configuration) we can skip this host
```
> we can skip this host

Host? Do you mean organization instead? "we can skip this organization"

john-westcott-iv replied:
Good catch, updated.

AlanCoding commented on awx/sso/backends.py:
```diff
@@ -327,31 +327,26 @@ def get_user(self, user_id):
-        user.save()
-        related.remove(user)
+        return True
+    return False
```
I don't see any case where `should_add` would not be a boolean, so it would be simpler to just `return should_add`. Before, this wasn't totally viable because it took these side-effect actions.

EDIT: you could also get rid of the `should_add` variable, with even greater condensing of code.

Also, what does this method do anymore? The docstring was probably incomplete before, but seems borderline incorrect now.

john-westcott-iv replied:
Method comment updated, `should_add` removed.

AlanCoding replied:
The comment I was thinking of was

> Hepler function to update m2m relationship based on LDAP group membership.

When I read that, I try to parse it into something that makes sense, and it just isn't happening.

AlanCoding replied:
just saw your update, resolved :+1:

john-westcott-iv committed:
Updating comments and remove unneeded variable

AlanCoding commented on awx/sso/backends.py:
```diff
@@ -423,3 +463,34 @@ def on_populate_user(sender, **kwargs):
+                    getattr(org, role).members.add(user)
+            elif user in getattr(org, role).members.all():
+                logger.debug("Removing LDAP user {} from {} in {}".format(user.username, role, org.name))
+                getattr(org, role).members.remove(user)
```
What you're doing is making the code easier to follow ultimately, which the primary data structures being `desired_org_states` and `desired_team_states`, that's good :+1: 

But you should allow yourself to modify the method signatures to reflect your improved code structure. This block, implemented in a separate method, would take `desired_org_states` and `user`. That could be tested independently to verify the correctness of the database queries and updates.

Name it like `update_database_to_LDAP_orgs` or `save_LDAP_desired_org_memberships`

AlanCoding commented on awx/sso/backends.py:
```diff
@@ -383,31 +381,73 @@ def on_populate_user(sender, **kwargs):
+    existing_org_names = list(Organization.objects.all().values_list('name', flat=True))
+    for org_name in org_map.keys():
+        if org_name not in existing_org_names:
+            Organization.objects.get_or_create(name=org_name)
```
If the org name isn't in `existing_org_names`, then deterministically, you create, right? Then I would opt for `create` instead of `get_or_create`, as it will be more performant and clear. If you are concerned about race conditions, this isn't the way to do things, you should do a try-except if you are really worried about it being created while in the loop so that you can later on make correct decisions.

AlanCoding commented on awx/sso/backends.py:
```diff
@@ -383,31 +381,73 @@ def on_populate_user(sender, **kwargs):
+        if team_name not in existing_team_names:
+            # Going a little inefficient here, we could prob get name/ids above and then set this by ID but I've got limited time r/n
+            #    and this code should only really be getting hit once if an LDAP config change adds a team with an org not built.
+            org, created = Organization.objects.get_or_create(name=team_opts['organization'])
```
Again, the use of `get_or_create` here doesn't make any sense to me, since all the prior checks just validated that the organization _already exists_ so this should only be a get, and never a create.

john-westcott-iv replied:
We can't technically guarantee that because there is no restriction that an organization tied to a team has to be in the org map. Consider an org map like:
```
{ "a", "b", "c"}
```
And a team map of:
```
{ "team1": { "org": "a" }, "team2": { "org": "d" } }
```
We would have never created "d" before this point.

AlanCoding replied:
That would hit `if team_opts['organization'] not in existing_org_names:` in the conditional right before this. So in that scenario, we would hit line 407 and line 413, and it doesn't make sense to me that both would be `get_or_create`. In both cases, there should be no ambiguity about whether it's getting or creating. It's one or the other for each. It may create and then get.

AlanCoding commented on awx/sso/backends.py:
```diff
@@ -383,31 +381,73 @@ def on_populate_user(sender, **kwargs):
+        if (
+            desired_org_states[org_name]['admin_role'] == None
+            and desired_org_states[org_name]['auditor_role'] == None
+            and desired_org_states[org_name]['member_role'] == None
```
I'd prefer to tighten this up, like `if all(desired_org_states[org_name][role_field] is None for role_field in roles):`

I also thought that the linter got you in trouble for `== None`.

where the definition of `roles = ['admin_role', 'auditor_role', 'member_role']` is shared with its use later on for saving the data.

john-westcott-iv committed:
Reverting checking of LDAP groups

The initial check performed case insensitive searches and the new method was case sensitive

The optimization of the new method is likely not going to contribute noticable slowness

john-westcott-iv committed:
Multiple enhancements

Extrapolating reconciliation of desired and actual states to a function

Converting heave prefect related methods to user focus for query optimization

Converting from get_or_create to simply create

Added memory calculations for query optimization

AlanCoding commented on awx/sso/backends.py:
```diff
@@ -423,3 +468,52 @@ def on_populate_user(sender, **kwargs):
+    reconsile_users_org_team_mappings(user, desired_org_states, desired_team_states, 'LDAP')
+
+
+def reconsile_users_org_team_mappings(user, desired_org_states, desired_team_states, source):
```
spelling of "reconsile"

AlanCoding commented on awx/sso/backends.py:
```diff
@@ -423,3 +468,52 @@ def on_populate_user(sender, **kwargs):
+    team_content_type = ContentType.objects.get_for_model(Team)
+
+    # users_roles is a flat list of IDs
+    users_roles = list(user.roles.filter(content_type__in=[org_content_type, team_content_type]).values_list('pk', flat=True))
```
Would be technically faster as `set`, and it looks like the code allows for that.

AlanCoding commented on awx/sso/backends.py:
```diff
@@ -423,3 +468,52 @@ def on_populate_user(sender, **kwargs):
         profile.ldap_dn = ldap_user.dn
         profile.save()
+
+    reconcile_users_org_team_mappings(user, desired_org_states, desired_team_states, 'LDAP')
```
A big non-obvious performance tweak I'd like to have is to handle the case that `desired_org_states` and `desired_team_states` are effectively nothing, in practice, `{}`

That could come in the form of not calling this method when those are empty, or, inside of that method we could bail early. Because the `ContentType.objects.get_for_model` has a query cost and `user_roles` is potentially costly in terms of the volume of return data, and we just don't want to run those when there's nothing to reconcile.

This might already be handled upstream of this, and if it is, you can disregard this comment.

AlanCoding commented on awx/sso/backends.py:
```diff
@@ -423,3 +468,52 @@ def on_populate_user(sender, **kwargs):
+                if role_name not in roles:
+                    roles.append(role_name)
+
+        # Get a set of named tuples for the org name plus all of the roles we got above
```
> org name

Could be org or team name.

AlanCoding commented on awx/sso/backends.py:
```diff
@@ -423,3 +468,52 @@ def on_populate_user(sender, **kwargs):
+                    roles.append(role_name)
+
+        # Get a set of named tuples for the org name plus all of the roles we got above
+        model_roles = getattr(model, 'objects').filter(name__in=desired_states.keys()).values_list(*(['name'] + roles), named=True)
```
There's a lot of indirection going on here, so I'd prefer to clean it up a bit for readability. Even after someone gets past those readability issues, it uses Django feature that most people won't be familiar with.

So to reduce indirection, I'd suggest

```python
model_roles = model.objects.filter(name__in=desired_states.keys()).values_list('name', *roles, named=True)
```

AlanCoding commented on awx/sso/backends.py:
```diff
@@ -423,3 +468,52 @@ def on_populate_user(sender, **kwargs):
+        for row in model_roles:
+            for role_name in roles:
+                desired_state = desired_states.get(row.name, {})
+                if desired_state[role_name] == None:
```
a standard python-ism is that we compare `is None` as opposed to the `==`. If you're worried that might affect your logic we can talk, but you can probably just drop that change in without any issues.

john-westcott-iv committed:
Fixing '== None' and better handeling of {} settings

AlanCoding commented on awx/sso/backends.py:
```diff
@@ -383,31 +386,73 @@ def on_populate_user(sender, **kwargs):
+    # Do the same for teams
+    existing_team_names = list(Team.objects.all().values_list('name', flat=True))
+    for team_name, team_opts in team_map.items():
+        if 'organization' not in team_opts:
```
Having read the prior organization loop, I think this should be `if not team_opts.get('organization'):`, to handle the case that the team mapping has an organization that's blank or null.

john-westcott-iv committed:
Changing to handle not only missing but null and empty organization in team map

AlanCoding reviewed (approved):
This work has matured very nicely. There are clearly delineated functional parts to this now, and I don't see anything that could easily be slower than what was there before.

jay-steurer commented:
Verified by QE
